leakcanary-android-core/src/main/java/leakcanary/internal/InternalLeakCanary.kt

```diff
@@ -106,6 +107,24 @@ internal object InternalLeakCanary : (Application) -> Unit, OnObjectRetainedList
 
   var resumedActivity: Activity? = null
 
+  private lateinit var heapDumpPrefs: SharedPreferences
+
+  internal var dumpEnabledInAboutScreen: Boolean
+    get() {
+      return if (::heapDumpPrefs.isInitialized) {
+        heapDumpPrefs
+                .getBoolean("AboutScreenDumpEnabled", true)
+      } else true
```

Comment:
I'm not sure we should return true as a default value here, instead we should probably block until the shared preferences are loaded. Which, granted, will mean IO / blocking on API 27 but heh. A simple way to do that is create a CountDownLatch(1),  call await here and countDown where we initialize shared prefs. Also, I'm not sure lateinit is volatile, if not we need to make this volatile.

Reply:
Would a better way out be by allowing disk read for shared prefs (temporarily) instead since blocking I/O can't be skipped?

```
val oldPolicy = StrictMode.allowThreadDiskReads()
heapDumpPrefs = application.getSharedPrefs("LeakCanaryHeapDumpPrefs", Context.MODE_PRIVATE)
StrictMode.setThreadPolicy(oldPolicy)
```

otherwise could go ahead with the countdown latch implementation where the latch is a scoped variable (during initialization), wdyt?

> Also, I'm not sure lateinit is volatile, if not we need to make this volatile.
Since the `InternalLeakCanary` object is a singleton (should also be volatile), I was assuming all it's mutable fields to be volatile implicitly, is that an incorrect understanding?

Reply:
> Since the InternalLeakCanary object is a singleton (should also be volatile), I was assuming all its mutable fields to be volatile implicitly, is that an incorrect understanding?

I'm not aware of the Kotlin compiler doing anything like that, let me know if you've seen otherwise.


That beeing said, yes `StrictMode.allowThreadDiskReads()` might be the most straightforward fix here.

Reply:
done the changes to allow disk reads

> I'm not aware of the Kotlin compiler doing anything like that, let me know if you've seen otherwise.

Bad assumption on my part, confirmed that the singleton instance is not volatile.